apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1/UrlSigner.PostPolicy.cs

```diff
@@ -0,0 +1,315 @@
+            /// </summary>
+            Key = 8,
+
+            /// <summary>
+            /// The allowed URL to which the client is redirected upon successful upload.
+            /// </summary>
+            SuccessActionRedirect = 9,
+
+            /// <summary>
+            /// The status code returned to the client upon successful upload if 
+            /// <see cref="SuccessActionRedirect"/> is not specified.
+            /// Allowed values are <see cref="HttpStatusCode.OK"/>, <see cref="HttpStatusCode.Created"/>,
+            /// <see cref="HttpStatusCode.NoContent"/>.
+            /// </summary>
+            SuccessActionStatus = 10,
+        }
+
+        /// <summary>
+        /// A custom element that can be added to a <see cref="PostPolicy"/>.
+        /// </summary>
+        public enum PostPolicyCustomElement
+        {
+            /// <summary>
+            /// Custom element used for requiering extra Google metadata to be included
+            /// in the post request. If the element name you provide is not prefixed by
+            /// x-goog-meta-, it will be prefixed with x-goog-meta- before being added to
+            /// the <see cref="PostPolicy"/>.
+            /// </summary>
+            GoogleMetadata = 0,
+
+            /// <summary>
+            /// Custom element used for requiering extra fields to be included in the post request.
+            /// </summary>
+            Custom = 1,
+        }
+
+        /// <summary>
+        /// Represents an unsigned post policy. It needs to be signed with <see cref="UrlSigner.Sign(PostPolicy, Options)"/> 
+        /// before it is used to enforced policies when posting objects via form.
+        /// </summary>
+        public sealed class PostPolicy
+        {
+            private static readonly IPostPolicyElement[] s_knownElements = new IPostPolicyElement[]
+            {
+                Acl.Element,
+                BucketName.Element,
+                CacheControl.Element,
+                ContentDisposition.Element,
+                ContentEncoding.Element,
+                ContentLength.Element,
+                ContentType.Element,
+                ObjectExpiry.Element,
+                Key.Element,
+                SuccessActionRedirect.Element,
+                SuccessActionStatus.Element
+            };
+
+            private static readonly ICustomElementFactory[] s_customElementFactories = new ICustomElementFactory[]
+            {
+                GoogleMetadataElementFactory.Default,
+                CustomElementFactory.Default
+            };
+
+            private readonly IDictionary<IPostPolicyElement, IPostPolicyCondition> _conditions =
+                new Dictionary<IPostPolicyElement, IPostPolicyCondition>();
+
+            private PostPolicy() { }
+
+            internal PostPolicy(PostPolicy other)
+            {
+                GaxPreconditions.CheckNotNull(other, nameof(other));
+                foreach (var condition in other._conditions)
```

Comment:
An alternative would be to initialize `_conditions` in each constructor, and make this one just use `new Dictionary<...>(other._conditions)`